beacon-chain/p2p/peers/scorers/data_column_rpc_request_scorer.go

```diff
@@ -0,0 +1,153 @@
+package scorers
+
+import (
+	"math"
+	"time"
+
+	"github.com/OffchainLabs/prysm/v6/beacon-chain/p2p/peers/peerdata"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/pkg/errors"
+)
+
+var _ Scorer = (*DataColumnRPCRequestScorer)(nil)
+
+const (
+	// DefaultDataColumnRPCRequestDecayInterval defines how often the decaying routine is called.
+	DefaultDataColumnRPCRequestDecayInterval = 30 * time.Second
+	// DefaultDataColumnRPCRequestDecay defines default number of requests that are to be subtracted
+	// from stats on each decay interval.
+	DefaultDataColumnRPCRequestDecay = uint64(10)
+	// DefaultDataColumnRPCRequestThreshold defines the maximum number of requests a peer can make
+	// before being considered bad.
+	DefaultDataColumnRPCRequestThreshold = uint64(100)
+	// DefaultDataColumnRPCRequestPenaltyFactor defines the penalty factor applied to request count.
+	DefaultDataColumnRPCRequestPenaltyFactor = float64(0.02)
+)
+
+// DataColumnRPCRequestScorer represents scoring service for tracking data column RPC requests.
+type DataColumnRPCRequestScorer struct {
+	config *DataColumnRPCRequestScorerConfig
+	store  *peerdata.Store
+}
+
+// DataColumnRPCRequestScorerConfig holds configuration parameters for data column RPC request scoring service.
+type DataColumnRPCRequestScorerConfig struct {
+	// DecayInterval defines how often stats should be decayed.
+	DecayInterval time.Duration
+	// Decay specifies number of requests subtracted from stats on each decay step.
+	Decay uint64
+	// Threshold defines maximum number of requests before peer is considered bad.
+	Threshold uint64
+	// PenaltyFactor defines multiplier applied to request count when calculating score.
+	PenaltyFactor float64
+}
+
+// newDataColumnRPCRequestScorer creates new scoring service for data column RPC requests.
+func newDataColumnRPCRequestScorer(store *peerdata.Store, config *DataColumnRPCRequestScorerConfig) *DataColumnRPCRequestScorer {
+	// Create a new config with default values
+	cfg := &DataColumnRPCRequestScorerConfig{
+		DecayInterval: DefaultDataColumnRPCRequestDecayInterval,
+		Decay:         DefaultDataColumnRPCRequestDecay,
+		Threshold:     DefaultDataColumnRPCRequestThreshold,
+		PenaltyFactor: DefaultDataColumnRPCRequestPenaltyFactor,
+	}
+
+	// Override with provided config values if present
+	if config != nil {
+		if config.DecayInterval != 0 {
+			cfg.DecayInterval = config.DecayInterval
+		}
+		if config.Decay != 0 {
+			cfg.Decay = config.Decay
+		}
+		if config.Threshold != 0 {
+			cfg.Threshold = config.Threshold
+		}
+		if config.PenaltyFactor != 0 {
+			cfg.PenaltyFactor = config.PenaltyFactor
+		}
+	}
```

Comment:
Not a direct comment on this PR, but more of a general note for, this pattern shows up in quite a few places, and we could consider replacing it with a generic helper like:

```
func override[T comparable](val T, def T) T {
	var zero T
	if val != zero {
		return val
	}
	return def
}
```

Reply:
You could handle this without defining a new generic function:`max(cfg.PenaltyFactor, config.PenaltyFactor)`